# Pair Programming ANOVA

1. En el ejercicio de hoy tendréis que hacer un ANOVA con vuestro datos y hacer una interpretación de los resultados.

In [1]:
import numpy as np
import pandas as pd
import random 

# Gráficos
# ------------------------------------------------------------------------------
import matplotlib.pyplot as plt
import seaborn as sns

# Estadísticos
# ------------------------------------------------------------------------------
import statsmodels.api as sm
from statsmodels.formula.api import ols # para realizar el ANOVA
from statsmodels.multivariate.manova import MANOVA
from sklearn.preprocessing import StandardScaler

plt.rcParams["figure.figsize"] = (10,8) 

In [2]:
df = pd.read_csv("datos/wellness_5.csv", index_col=0)
df.head(2)

FRUITS_VEGGIES DAILY_STRESS  CORE_CIRCLE  SOCIAL_NETWORK  BMI_RANGE  \
0               3            2            5               5          1   
1               2            3            3              10          2   

   DAILY_STEPS  SLEEP_HOURS  SUFFICIENT_INCOME  TIME_FOR_PASSION  \
0            5            7                  1                 0   
1            5            8                  2                 2   

   WEEKLY_MEDITATION       AGE  GENDER  WORK_LIFE_BALANCE_SCORE  \
0                  5  36 to 50  Female                    609.5   
1                  6  36 to 50  Female                    655.6   

   WORK_LIFE_BALANCE_SCORE_RAIZ  
0                     24.688054  
1                     25.604687

In [3]:
df.columns

Index(['FRUITS_VEGGIES', 'DAILY_STRESS', 'CORE_CIRCLE', 'SOCIAL_NETWORK',
       'BMI_RANGE', 'DAILY_STEPS', 'SLEEP_HOURS', 'SUFFICIENT_INCOME',
       'TIME_FOR_PASSION', 'WEEKLY_MEDITATION', 'AGE', 'GENDER',
       'WORK_LIFE_BALANCE_SCORE', 'WORK_LIFE_BALANCE_SCORE_RAIZ'],
      dtype='object')

In [12]:
df.head()

FRUITS_VEGGIES DAILY_STRESS  CORE_CIRCLE  SOCIAL_NETWORK  BMI_RANGE  \
0               3            2            5               5          1   
1               2            3            3              10          2   
2               2            3            4              10          2   
3               3            3            3               7          2   
4               5            1            3               4          2   

   DAILY_STEPS  SLEEP_HOURS  SUFFICIENT_INCOME  TIME_FOR_PASSION  \
0            5            7                  1                 0   
1            5            8                  2                 2   
2            4            8                  2                 8   
3            5            5                  1                 2   
4            5            7                  2                 1   

   WEEKLY_MEDITATION         AGE  GENDER  WORK_LIFE_BALANCE_SCORE  \
0                  5    36 to 50  Female                    609.5   
1                  6    36 to 50  Female                    655.6   
2                  3    36 to 50  Female                    631.6   
3                  0  51 or more  Female                    622.7   
4                  5  51 or more  Female                    663.9   

   WORK_LIFE_BALANCE_SCORE_RAIZ  
0                     24.688054  
1                     25.604687  
2                     25.131653  
3                     24.953958  
4                     25.766257

In [5]:
df = df.drop(["WORK_LIFE_BALANCE_SCORE"],axis=1, inplace= True)

In [6]:
df.head()

AttributeError: 'NoneType' object has no attribute 'head'

In [14]:
lm = ols('WORK_LIFE_BALANCE_SCORE_RAIZ ~ FRUITS_VEGGIES + DAILY_STRESS + CORE_CIRCLE + SOCIAL_NETWORK + BMI_RANGE + DAILY_STEPS + SLEEP_HOURS +SUFFICIENT_INCOME + TIME_FOR_PASSION + WEEKLY_MEDITATION + AGE + GENDER' , data=df).fit()
sm.stats.anova_lm(lm)
#Dado que la finalidad de hacer una ANOVA es saber qué variables debemos descartar, no tendría sentido hacerla 
# multiplicando las variables entre sí por todas las interaciones que eso conllevaría.

df       sum_sq      mean_sq             F  \
DAILY_STRESS           6.0  1602.339727   267.056621   1845.855646   
AGE                    3.0   172.454369    57.484790    397.326316   
GENDER                 1.0    69.035525    69.035525    477.163282   
FRUITS_VEGGIES         1.0  1836.233732  1836.233732  12691.774448   
CORE_CIRCLE            1.0  1952.765005  1952.765005  13497.221279   
SOCIAL_NETWORK         1.0   847.002147   847.002147   5854.352864   
BMI_RANGE              1.0   530.972859   530.972859   3670.005430   
DAILY_STEPS            1.0   477.978945   477.978945   3303.719374   
SLEEP_HOURS            1.0   119.929507   119.929507    828.934914   
SUFFICIENT_INCOME      1.0   566.333799   566.333799   3914.414988   
TIME_FOR_PASSION       1.0   952.501131   952.501131   6583.546148   
WEEKLY_MEDITATION      1.0   380.704155   380.704155   2631.370496   
Residual           15266.0  2208.670211     0.144679           NaN   

                          PR(>F)  
DAILY_STRESS        0.000000e+00  
AGE                1.574655e-248  
GENDER             3.467668e-104  
FRUITS_VEGGIES      0.000000e+00  
CORE_CIRCLE         0.000000e+00  
SOCIAL_NETWORK      0.000000e+00  
BMI_RANGE           0.000000e+00  
DAILY_STEPS         0.000000e+00  
SLEEP_HOURS        1.478000e-177  
SUFFICIENT_INCOME   0.000000e+00  
TIME_FOR_PASSION    0.000000e+00  
WEEKLY_MEDITATION   0.000000e+00  
Residual                     NaN

* df: son los grados de libertad, el número de observaciones en los datos que pueden variar libremente (en las variables numéricas siempre será 1)
* sum_sq: es una medida de variación o desviación de la media. Se calcula como una suma de los cuadrados de las diferencias con respecto a la media
* mean_sq: resultado de dividir la suma de cuadrados entre el nº de grados de libertad
* PR(>F): es el p-valor. Si es <o = 0,05, rechazamos la H0 y concluimos que esa variable si influye en la variable respuesta
Si p-valor  > 0, no contamos con evidencias suficientes para rechazar la H0. Diremos que esa variable no tiene efecto sobre el precio de las casas

In [15]:
lm.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                                 OLS Regression Results                                 
========================================================================================
Dep. Variable:     WORK_LIFE_BALANCE_SCORE_RAIZ   R-squared:                       0.811
Model:                                      OLS   Adj. R-squared:                  0.811
Method:                           Least Squares   F-statistic:                     3459.
Date:                          Thu, 26 Jan 2023   Prob (F-statistic):               0.00
Time:                                  09:46:13   Log-Likelihood:                -6904.2
No. Observations:                         15286   AIC:                         1.385e+04
Df Residuals:                             15266   BIC:                         1.400e+04
Df Model:                                    19                                         
Covariance Type:                      nonrobust                                         
==========================================================================================
                             coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------------------
Intercept                 23.3327      0.031    743.715      0.000      23.271      23.394
DAILY_STRESS[T.1]         -0.0632      0.017     -3.747      0.000      -0.096      -0.030
DAILY_STRESS[T.1/1/00]     0.1138      0.381      0.299      0.765      -0.633       0.860
DAILY_STRESS[T.2]         -0.1617      0.016     -9.818      0.000      -0.194      -0.129
DAILY_STRESS[T.3]         -0.2587      0.016    -15.886      0.000      -0.291      -0.227
DAILY_STRESS[T.4]         -0.3857      0.017    -22.871      0.000      -0.419      -0.353
DAILY_STRESS[T.5]         -0.5120      0.018    -28.883      0.000      -0.547      -0.477
AGE[T.36 to 50]            0.0548      0.008      7.097      0.000       0.040       0.070
AGE[T.51 or more]          0.1340      0.009     15.354      0.000       0.117       0.151
AGE[T.Less than 20]       -0.0823      0.011     -7.819      0.000      -0.103      -0.062
GENDER[T.Male]            -0.0596      0.007     -9.096      0.000      -0.072      -0.047
FRUITS_VEGGIES             0.1045      0.002     44.727      0.000       0.100       0.109
CORE_CIRCLE                0.0751      0.001     63.494      0.000       0.073       0.077
SOCIAL_NETWORK             0.0603      0.001     54.852      0.000       0.058       0.062
BMI_RANGE                 -0.3300      0.007    -50.470      0.000      -0.343      -0.317
DAILY_STEPS                0.0523      0.001     45.738      0.000       0.050       0.055
SLEEP_HOURS                0.0486      0.003     18.273      0.000       0.043       0.054
SUFFICIENT_INCOME          0.4515      0.007     62.461      0.000       0.437       0.466
TIME_FOR_PASSION           0.0890      0.001     73.872      0.000       0.087       0.091
WEEKLY_MEDITATION          0.0565      0.001     51.297      0.000       0.054       0.059
==============================================================================
Omnibus:                        1.647   Durbin-Watson:                   1.943
Prob(Omnibus):                  0.439   Jarque-Bera (JB):                1.671
Skew:                          -0.023   Prob(JB):                        0.434
Kurtosis:                       2.978   Cond. No.                     1.88e+03
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The condition number is large, 1.88e+03. This might indicate that there are
strong multicollinearity or other numerical problems.
"""

Lo primero que tenemos que observar al hacer un summary es el p-valor, y podremos concluir que siempre que esté por debajo de 0,05; esa variable influye en los valores de nuestra variable respuesta. Cuando el valor sea mayor que 0,05, podremos descartar la variable por no ser importante en la tendencia de nuestra variable respuesta.

En nuestro caso vemos que la única variable con un p-valor mayor de 0,05 es "DAILY_STRESS[T.1/1/00]" de manera que podríamos concluir que esa variable no influye en nuestra variable respuesta.

El coeficiente proporciona la tasa de cambio y lo que nos interesa es ver el signo que presenta. Por ejemplo, cuanto mayor es el tiempo dedicado a actividades que nos gustan, la ingesta de fruta y verdura, los pasos diarios o el círculo social, horas de sueño, etc; mayor es la sensación de bienestar. 
En el caso del índice de masa corporal sucede al contrario, cuanto mayor es el BMI menor es la sensación de bienestar.

El R2 nos indica que porcentaje del cambio en la variable respuesta se debe a las variables predictoras, en nuestro caso es del 81%.